flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java

```diff
@@ -18,176 +18,905 @@
+                        .testCase(FLOAT(), 9234567891.12, 9234567891.12d)
+                        .testCase(DOUBLE(), 123.4567890, 123.456789d)
+                        .testCase(DOUBLE(), 1239234567891.1234567891234, 1.2392345678911235E12d)
+                        // Not supported - no fix
+                        // DATE
+                        // TIME
+                        // TIMESTAMP
+                        // TIMESTAMP_WITH_TIME_ZONE
+                        // TIMESTAMP_LTZ
+                        // INTERVAL(YEAR())
+                        // INTERVAL(DAY())
+                        // ARRAY
+                        // MULTISET
+                        // MAP
+                        // ROW
+                        // RAW
+                        .build(),
+                CastTestSpecBuilder.test(DATE(), "To DATE")
+                        .testCase(CHAR(3), "foo", null)
+                        .testCase(VARCHAR(5), "Flink", null)
+                        // https://issues.apache.org/jira/browse/FLINK-24421 Bug
+                        .testCase(STRING(), "123", LocalDate.of(123, 1, 1))
+                        .testCase(STRING(), "2021-09-27", LocalDate.of(2021, 9, 27))
+                        .testCase(STRING(), "2021/09/27", null)
+                        // Not supported - no fix
+                        // BOOLEAN
+                        // BINARY
+                        // VARBINARY
+                        // BYTES
+                        // DECIMAL
+                        // TINYINT
+                        // SMALLINT
+                        // INT
+                        // BIGINT
+                        // FLOAT
+                        // DOUBLE
+                        .testCase(DATE(), LocalDate.parse("2021-09-24"), LocalDate.of(2021, 9, 24))
+                        // Not supported - no fix
+                        // TIME
+                        .testCase(
+                                TIMESTAMP(),
+                                LocalDateTime.parse("2021-09-24T12:34:56.123456"),
+                                LocalDate.of(2021, 9, 24))
+                        .testCase(
+                                TIMESTAMP(4),
+                                LocalDateTime.parse("2021-09-24T12:34:56.123456"),
+                                LocalDate.of(2021, 9, 24))
+                        // Not supported currently
+                        // TIMESTAMP_WITH_TIME_ZONE
+                        //
+                        // https://issues.apache.org/jira/browse/FLINK-24422 - Accept only Instant
+                        .testCase(
+                                TIMESTAMP_LTZ(),
+                                LocalDateTime.parse("2021-09-24T12:34:56.123456"),
+                                LocalDate.of(2021, 9, 24))
+                        .testCase(
+                                TIMESTAMP_LTZ(),
+                                LocalDateTime.parse("2021-09-24T22:34:56.123456")
```

Comment:
how does the instant look like after this? we should design the test in a way that the instant shows `24` but the resulting `LocalDate` is 25 due to the session timezone

Reply:
or reference `TEST_TZ` instead of `5`

Reply:
I am a bit confused, isn't the case currently?

Reply:
I ended up using both a TEST_TZ and a TEST_OFFSET which are different, so the OFFSET is applied when creating source timestamps and the TEST_TZ to simulate the session timezone set by a user.